test/unit/permutations/test_permutations.cpp

```diff
@@ -0,0 +1,267 @@
+
+    // column 7
+    {9,  TileElementIndex(1, 1), 11},
+    {9,  TileElementIndex(2, 1), 12},
+    {10, TileElementIndex(0, 1), 13},
+  };
+  // clang-format on
+
+  for (auto [i_tile, i_el, val] : expected_entries) {
+    EXPECT_EQ(out_tiles[i_tile](i_el), val);
+  }
+}
+
+// The following row permutations are tested:
+//
+// ┌──────────────────────┐    ┌──────────────────────┐
+// │in                    │    │ out                  │
+// │     ****** a         │    │                      │
+// │                      │    │                      │
+// │                      │    │                      │
+// │     ****** b         │    │                      │
+// │                      │    │         c ******     │
+// │                      │    │         b ******     │
+// │     ****** c         │    │         a ******     │
+// │                      │    │                      │
+// │                      │    │                      │
+// └──────────────────────┘    └──────────────────────┘
+// where the (i, j) element of `in` is `i + j`
+//
+template <Device D, class T>
+void testApplyRowPermutations(SizeType n, SizeType nb) {
+  using dlaf::matrix::test::createTile;
+  using dlaf::matrix::test::set;
+
+  auto [distr, in_tiles, out_tiles] = setupMatricesForPermutations<T>(n, nb);
+
+  GlobalElementIndex begin(7, 3);
+  GlobalElementSize sz(3, 6);
+  SizeType in_offset = 2;
+  std::vector<SizeType> perm_arr{8, 4, 1};
+  dlaf::permutations::internal::applyPermutations<T, Coord::Row>(begin, sz, in_offset, distr,
+                                                                 perm_arr.data(), in_tiles, out_tiles);
+
+  // Expected entries in `out_tiles`
+  //
+  // (tile index in `out_tiles`, index of element in tile, value of element)
+  // clang-format off
+  std::vector<std::tuple<std::size_t, TileElementIndex, T>> expected_entries {
+    // row index 7
+    {6,  TileElementIndex(1, 0), 10},
+    {6,  TileElementIndex(1, 1), 11},
+    {6,  TileElementIndex(1, 2), 12},
+    {10, TileElementIndex(1, 0), 13},
+    {10, TileElementIndex(1, 1), 14},
+    {10, TileElementIndex(1, 2), 15},
+
+    // row index 8
+    {6,  TileElementIndex(2, 0), 6 },
+    {6,  TileElementIndex(2, 1), 7 },
+    {6,  TileElementIndex(2, 2), 8 },
+    {10, TileElementIndex(2, 0), 9 },
+    {10, TileElementIndex(2, 1), 10},
+    {10, TileElementIndex(2, 2), 11},
+
+    // row index 9
+    {7,  TileElementIndex(0, 0), 3},
+    {7,  TileElementIndex(0, 1), 4},
+    {7,  TileElementIndex(0, 2), 5},
+    {11, TileElementIndex(0, 0), 6},
+    {11, TileElementIndex(0, 1), 7},
+    {11, TileElementIndex(0, 2), 8},
+
+  };
+  // clang-format on
+
+  for (auto [i_tile, i_el, val] : expected_entries) {
+    EXPECT_EQ(out_tiles[i_tile](i_el), val);
+  }
+}
+
+// Permute columns or rows in reverse order.
+// Each column or row of the input matrix has its index as a value.
+template <Backend B, Device D, class T, Coord C>
+void testPermutations(SizeType n, SizeType nb) {
```

Comment:
Could it be useful to test this with non-square matrices?

Reply:
In principle, yes, it should work for non-square matrices as well. I haven't tested non-square matrices because the current use-case within the tridiagonal solver is square matrices.